lib/Runtime/Library/EngineInterfaceObject.cpp

```diff
@@ -317,30 +317,43 @@ namespace Js
+        AssertOrFailFast((callInfo.Count == 3 || callInfo.Count == 4) && JavascriptFunction::Is(args[1]) && JavascriptString::Is(args[2]));
+
+        JavascriptFunction *func = JavascriptFunction::UnsafeFromVar(args[1]);
+        JavascriptString *methodName = JavascriptString::UnsafeFromVar(args[2]);
+
+        func->GetFunctionProxy()->SetIsPublicLibraryCode();
+
+        // use GetSz rather than GetString because we use wcsrchr below, which expects a null-terminated string
+        const char16 *methodNameBuf = methodName->GetSz();
+        charcount_t methodNameLength = methodName->GetLength();
+        const char16 *shortName = wcsrchr(methodNameBuf, _u('.'));
+        charcount_t shortNameOffset = 0;
+        if (shortName != nullptr)
         {
-            JavascriptFunction* func = JavascriptFunction::FromVar(args.Values[1]);
-            func->GetFunctionProxy()->SetIsPublicLibraryCode();
+            shortName++;
+            shortNameOffset = static_cast<charcount_t>(shortName - methodNameBuf);
+        }
 
-            if (callInfo.Count >= 3 && JavascriptString::Is(args.Values[2]))
-            {
-                JavascriptString* customFunctionName = JavascriptString::FromVar(args.Values[2]);
-                // tagPublicFunction("Intl.Collator", Collator); in Intl.js calls TagPublicLibraryCode the expected name is Collator so we need to calculate the offset
-                const char16 * shortName = wcsrchr(customFunctionName->GetString(), _u('.'));
-                uint shortNameOffset = 0;
-                if (shortName != nullptr)
-                {
-                    // JavascriptString length is bounded by uint max
-                    shortName++;
-                    shortNameOffset = static_cast<uint>(shortName - customFunctionName->GetString());
-                }
-                func->GetFunctionProxy()->EnsureDeserialized()->SetDisplayName(customFunctionName->GetString(), customFunctionName->GetLength(), shortNameOffset);
-            }
+        func->GetFunctionProxy()->EnsureDeserialized()->SetDisplayName(methodNameBuf, methodNameLength, shortNameOffset);
 
-            return func;
+        bool creatingConstructor = true;
+        if (callInfo.Count == 4)
+        {
+            AssertOrFailFast(JavascriptBoolean::Is(args[3]));
+            creatingConstructor = JavascriptBoolean::UnsafeFromVar(args[3])->GetValue();
         }
 
-        return scriptContext->GetLibrary()->GetUndefined();
+        if (!creatingConstructor)
+        {
+            FunctionInfo *info = func->GetFunctionInfo();
+            info->SetAttributes((FunctionInfo::Attributes) (info->GetAttributes() | FunctionInfo::Attributes::ErrorOnNew));
+
+            AssertOrFailFast(func->GetDynamicType()->GetTypeHandler()->IsDeferredTypeHandler());
+            DynamicTypeHandler::SetInstanceTypeHandler(func, scriptContext->GetLibrary()->GetDeferredFunctionTypeHandlerNoPrototype());
```

Comment:
@pleath @boingoing Thoughts on this way of making sure these builtins don't have prototypes, rather than forcefully deleting the prototype property?

Reply:
Per the piece I did on adding length properties - this type seems correct for this BUT - the tt-debug errors you're getting look like a problem I had with my initial attempts at changing function types to add lengths so there's a chance that this change caused the TTDebug test fails - I got that particular test fail when my types were wrong such that accessing a function caused it's type to transform.

Reply:
@jackhorton I added that type in #5405 and was now planning on renaming it in #5610 (as other function types say "prototype" in their name if they have one and don't mention it if they don't whereas they explicitly mention if they have a length - which this does)

Just an FYI that if 5610 is merged before this and this used here will need to rename it here (unless I drop the name change from 5610)

Reply:
Yeah I was going to wait to merge this until we got 5610 and fromEntries merged so that you dont have to keep rebasing/re-bytecode-genning on top of all of our changes. I wanted to wait for some more opinion on 5610 but I may just merge it today.